### ISSUE TITLE
Kanban board

### ISSUE THREAD

**Issue by HLFH** [gogits/gogs#3716]
Hi,

It would be great to add a Kanban board feature to Gogs. GitHub and GitLab are doing it.

Thanks,
HLFH

**Comment by skrzepto** [comment]
Do we force the generic 4 columns (backlog, ready, wip, review) or do we allow the admin to add any column they wish?

**Comment by HLFH** [comment]
Yes at least, we should be able to have 4 columns (backlog, next, in progress, released). Wekan allows also to add any column we wish. IMHO, 4 columns by default would be interesting. But we should be able to change the number and the names of the columns.

**Comment by skrzepto** [comment]
> But we should be able to change the number and the names of the columns.

Could you elaborate what you mean by number? Do you mean the position of the column or the amount of columns?

**Comment by HLFH** [comment]
The amount of columns. And in fact, on [Wekan](https://github.com/wekan/wekan), the position can also be changed just by moving the mouse. But I don't expect something big for the first possible implementation of this feature within Gogs.

**Comment by dasyrw9hfh** [comment]
Having a kanban sounds like unnecessary bloat to me...

**Comment by skrzepto** [comment]
@dasyrw9hfh 
> Having a kanban sounds like unnecessary bloat to me..
Depends who you are talking with. In my opinion it would be nice feature to implement.

My thoughts to satisfy both sides would be  to leave it off default for new repos and the repo admin would have to tick a box to enable the kanboard feature. Thoughts?

**Comment by skrzepto** [comment]
One would argue that kanboard is implemented in https://kanboard.net/plugin/gogs-webhook

**Comment by genofire** [comment]
the column could easy be a tag on the issue.
On this way the name and order of the columns must only be stored on the project.

**Comment by skrzepto** [comment]
@genofire thats assuming if we focus on one board only? Should there be the ability to have multiple boards per repo or just one?

**Comment by genofire** [comment]
for me, one per repo is enought -> but github could use multiple, so for other users this could be more interesting

So the kanban tag on issue could be something like `boardId_columnname` or `boardname_columnname`. (another option would be two tags).

My favour would be that the columnname is in the tag name.

* Never mind the a renameing of the column would destroy the relationship. (the issues has not anything todo with the new column)
* Or we have to handle a renaming of the column by renaming all tags on all issues in a repo.

**Comment by 0xbaadf00d** [comment]
If Kanban could be disabled (like PR or Wiki) it could be nice. Kabban board is not necessary needed on all repos.

**Comment by skrzepto** [comment]
@genofire I agree one repo would be enough. And I think a unique ID to refer to the column would make sense for O(1) rename since the board_column_name can change while the uuid stays the same.


> If Kanban tab could be disabled (like PR or Wiki) it could be nice. Kanban board is not necessary needed on all repos.

@0xbaadf00d 100% agree, I started working on a initial implementation and that was the first thing I did.

**Comment by genofire** [comment]
If you need help ....

**Comment by skrzepto** [comment]
@genofire I think for the most part the backend code is close to finish so next step is UI/UX which is not my strong suit. 

Regarding the backend

Just have a question about how the issues should be returned in the front end. Example

----
Option 1:
Should we return an array of all the issues in JSON format and have JS iterate the list of issues and place them on the board

Option 2:
Or have the backend do the ordering and return {'column1': issue_1, 'column2': 'issue_2, etc...}

----

I'm leaning on Option 1 because it offloads the workload to the user but thats just a personal opinion. Thoughts?

**Comment by genofire** [comment]
I like to handel it on client side, for drag and drop it would be the best way i thing. (Option 1)

**Comment by mhtsbt** [comment]
If anyone is interessted in working on this, feel free to contact me. I started working (https://github.com/mhtsbt/GogsBoard) on a gogs-powered kanban board. It's a seperate application that connects through the api:

**Comment by lunny** [comment]
@mhtsbt could you also send a PR to Gitea? I think many maintainers of Gitea could review your code.

**Comment by RcrdBrt** [comment]
Ping?

**Comment by sebastianmacias** [comment]
+1 Kanban implementatio

**Comment by kirvesp** [comment]
As I see @mhtsbt ends his work with gogs canban project. Does anyone work on this in current time?

I would be glad to offer my help with this.

**Comment by kstan79** [comment]
Hi, is this plugin merged into master?

**Comment by RcrdBrt** [comment]
Nope, although it would be a very nice addition.

**Comment by BruinY** [comment]
i want

**Comment by elmbrent** [comment]
This would be great!

**Comment by kbaegis** [comment]
+1

**Comment by zlbabe** [comment]
Ping !

**Comment by Xsmael** [comment]
any update on this ?

**Comment by ckotthaus** [comment]
+1

**Comment by LucasManerich** [comment]
+1

**Comment by minias** [comment]
+1

**Comment by gemurdock** [comment]
I think a lot of people would love something like this. I know I use the Kanban board for every project that is not simple.

**Comment by fc-dev** [comment]
+1

**Comment by k9t9** [comment]
Is there any update on this feature ( 5 years on the pipe now ) ?

**Comment by Tha14** [comment]
Any update on this? It could really help with development.

**Comment by digitalextremist** [comment]
This is pretty much the only thing holding me back from migrating from `GitLab`

**Comment by genofire** [comment]
PS: gitea and forgejo has it already, i do not know how long, but here is a PR with screenshots of bugfixes there:
https://github.com/go-gitea/gitea/pull/13437

**Comment by Natahan-Tatan** [comment]
There is some advancement on that feature? 

Gogs is a very good light git server and is a great alternative to the very heavy Gitlab, too big and with too unnecessary features for tiny team. 
The ability to organize issues in a kanban is a great feature that hit needs of a wide range of small team or solo dev.
